### PR TITLE
samplerParameters sample

### PR DESCRIPTION
**Pull request by kainino0x**
Preview: https://kai.graphics/webgpu-samples/samples/samplerParameters
I'd like to make sure the sample is comprehensible before reading the code, so please check out the preview before reviewing.

### CODE REVIEW

File: src/sample/samplerParameters/main.ts [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+    format: presentationFormat,
+    alphaMode: 'premultiplied',
+  });
+
+  //
+  // Initialize test texture
+  //
+
+  // Set up a texture with 4 mip levels, each containing a checkerboard with 1x1
+  // pixels (such that the checkerboards do NOT line up).
```
Comment (lokokung) [inline src/sample/samplerParameters/main.ts:136]: Hmm not sure what you mean by "NOT line up", but what I thought it meant was like whether the (0,0) pixel is black or colored? But looking at the rendered textures and mips, it looks like they are actually lined up by that definition? So not sure what this means, maybe a bit more clarification could be helpful?

File: src/sample/samplerParameters/main.ts [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    [255, 193, 7, 255],
+    [216, 27, 96, 255],
+  ];
+  for (let mipLevel = 0; mipLevel < kTextureMipLevels; ++mipLevel) {
+    const size = 2 ** (kTextureMipLevels - mipLevel); // 16, 8, 4, 2
+    const data = new Uint8Array(size * size * 4);
+    for (let y = 0; y < size; ++y) {
+      for (let x = 0; x < size; ++x) {
+        data.set(
+          (x + y) % 2 ? kColorForLevel[mipLevel] : [0, 0, 0, 255],
```
Comment (lokokung) [inline src/sample/samplerParameters/main.ts:163]: Regarding the comment above about mips lining up, I think if you make this:
```
(x + y + mipLevel) % 2 ? kColorForLevel[mipLevel] : [0, 0, 0, 255]
```
We can achieve that.

File: src/sample/samplerParameters/main.ts [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+      {
+        name: './showTexture.wgsl',
+        contents: showTextureWGSL,
+        editable: true,
+      },
+    ],
+    filename: __filename,
+  });
+
+export default HelloTriangle;
```
Comment (lokokung) [inline src/sample/samplerParameters/main.ts:340]: nit. Rename before adding into the actual samples repo!

**Review by lokokung — COMMENTED** [review]
FIrst pass for now, I saw that you might be trying to add UV scrolling so will do another pass when that is in!

File: src/sample/samplerParameters/main.ts (hunk 1 above)
Comment (kainino0x) [reply src/sample/samplerParameters/main.ts:185]: What I meant was the checkerboards will all be different sizes, not like a real mipmap where you would have each one look like a lower-res version of the previous one. Will clarify

File: src/sample/samplerParameters/main.ts [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    device,
+    format: presentationFormat,
+    alphaMode: 'premultiplied',
+  });
+
+  //
+  // Initialize test texture
+  //
+
+  // Set up a texture with 4 mip levels, each containing a differently
```
Comment (lokokung) [inline src/sample/samplerParameters/main.ts:184]: nit. different

File: src/sample/samplerParameters/main.ts [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    };
+
+    const pass = commandEncoder.beginRenderPass(renderPassDescriptor);
+    // Draw test squares
+    pass.setPipeline(texturedSquarePipeline);
+    for (let i = 0; i < kViewportGridSize ** 2 - 1; ++i) {
+      const vpX = kViewportGridStride * (i % kViewportGridSize) + 1;
+      const vpY = kViewportGridStride * Math.floor(i / kViewportGridSize) + 1;
+      pass.setViewport(vpX, vpY, kViewportSize, kViewportSize, 0, 1);
+      pass.setBindGroup(0, bindGroup);
```
Comment (lokokung) [inline src/sample/samplerParameters/main.ts:342]: I think this could go outside of the for loop?

**Review by lokokung — APPROVED** [review]
LGTM!

**Review by lokokung — APPROVED** [review]
LGTM!

**Review by austinEng — APPROVED** [review]

**Pull request merged**